src/libspark/keys.cpp

```diff
@@ -12,6 +13,19 @@ SpendKey::SpendKey(const Params* params) {
 	this->r.randomize();
 }
 
+SpendKey::SpendKey(const Params* params, const Scalar& r_) {
+    this->params = params;
+    this->r = r_;
+    std::vector<unsigned char> data;
+    r.serialize(data.data());
+    std::vector<unsigned char> result(CSHA512().OUTPUT_SIZE);
+    CHash512 hash512;
+    hash512.Write(data.data(), data.size()).Finalize(&result[0]);
+
+    this->s1.memberFromSeed(&result[0]);
+    this->s2.memberFromSeed(&result[32]);
```

Comment:
If desired, it is also safe (with appropriate domain separation and any other KDF considerations) to hash `r` to obtain `s2`, and then hash `s2` to obtain `s1`.